pkg/upgrade/jiva.go

```diff
@@ -0,0 +1,243 @@
+	logLevel           int32
+}
+
+// Jiva Data-plane Upgrade Job instantiator
+func InstantiateJivaUpgrade(openebsNs string, toVersion string, menifestFile string) {
+	k, err := client.NewK8sClient("")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error creating k8s client")
+		return
+	}
+
+	// If manifest Files is provided, apply the file to create a new upgrade-job
+	if menifestFile != "" {
+		yamlFile, err := yamlToJobSpec(menifestFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error in Job: %s", err)
+		}
+		k.CreateBatchJob(yamlFile)
+		return
+	}
+
+	volNames, fromVersion, desiredVersion, err := GetJivaVolumes(k)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	if toVersion == "" {
+		if desiredVersion != fromVersion {
+			// Mark it as toVersion
+			toVersion = desiredVersion
+		} else {
+			// TODO: Upgrade version to latest available version for Jiva volumes
+			fmt.Println("Fetching latest version from the remote...")
+			latVer, err := getLatestJivaVersion()
+			if err != nil {
+				log.Fatal("Error fetching latest version: ", err)
+			}
+			toVersion = latVer
+		}
+	}
+
+	if openebsNs == "" {
+		fmt.Println(`No Namespace Provided, using "default" as a namespace`)
+		openebsNs = "default"
+	}
+
+	cfg := jivaUpdateConfig{
+		fromVersion:        fromVersion,
+		toVersion:          toVersion,
+		namespace:          openebsNs,
+		pvNames:            volNames,
+		serviceAccountName: "jiva-operator",
+		backOffLimit:       4,
+		logLevel:           4,
+	}
+
+	jobSpec := GetJivaBatchJob(&cfg)
+	k.CreateBatchJob(jobSpec)
+}
+
+// GetJivaVolumes returns the Jiva volumes list and current version
+func GetJivaVolumes(k *client.K8sClient) ([]string, string, string, error) {
+	// 1. Fetch all jivavolumes CRs in all namespaces
+	_, jvMap, err := k.GetJVs(nil, util.Map, "", util.MapOptions{Key: util.Name})
+	if err != nil {
+		return nil, "", "", fmt.Errorf("err getting jiva volumes: %s", err.Error())
+	}
+
+	var jivaList *corev1.PersistentVolumeList
+	//2. Get Jiva Persistent volumes
+	jivaList, err = k.GetPvByCasType([]string{"jiva"}, "")
+	if err != nil {
+		return nil, "", "", fmt.Errorf("err getting jiva volumes: %s", err.Error())
+	}
+
+	var volumeNames []string
+	var version, desiredVersion string
+
+	//3. Write-out names, versions and desired-versions
+	for _, pv := range jivaList.Items {
+		volumeNames = append(volumeNames, pv.Name)
+		if v, ok := jvMap[pv.Name]; ok && len(version) == 0 {
+			version = v.VersionDetails.Status.Current
+			desiredVersion = v.VersionDetails.Desired
+		}
+	}
+
+	//4. Check for zero jiva-volumes
+	if len(version) == 0 || len(volumeNames) == 0 {
+		return volumeNames, version, desiredVersion, fmt.Errorf("no jiva volumes found")
+	}
+
+	return volumeNames, version, desiredVersion, nil
+}
+
+// GetJivaBatchJob returns the Jiva Batch Specifications
+func GetJivaBatchJob(cfg *jivaUpdateConfig) *batchV1.Job {
+	jobSpec := &batchV1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "jiva-volume-upgrade",
```

Comment:
We need to make this configurable for multiple jobs or use `GeneratedName` field